src/mixins/java/org/spongepowered/common/mixin/core/commands/CommandsMixin.java

```diff
@@ -208,57 +208,57 @@
-                        this.impl$playerNodeCache.get(e),
-                        !hasCustomSuggestionsAlready
-                );
-                if (!this.impl$getChildrenFromNode(commandNode).isEmpty()) {
-                    this.shadow$fillUsableCommands(commandNode, finalCommandNode, sourceButTyped, commandNodeToSuggestionNode);
-                }
-            }
-
-            if (shouldContinue) {
-                // If we've already created the node, then just attach what we already have.
-                final ServerPlayer e = (ServerPlayer) sourceButTyped.getEntity();
-                final List<CommandNode<SharedSuggestionProvider>> suggestionProviderCommandNode = this.impl$playerNodeCache.get(e).get(commandNode);
-                if (suggestionProviderCommandNode != null) {
-                    shouldContinue = false;
-                    boolean hasCustomSuggestionsAlready = CommandUtil.checkForCustomSuggestions(rootSuggestion);
-                    for (final CommandNode<SharedSuggestionProvider> node : suggestionProviderCommandNode) {
-                        // If we have custom suggestions, we need to limit it to one node, otherwise we trigger a bug
-                        // in the client where it'll send more than one custom suggestion request - which is fine, except
-                        // the client will then ignore all but one of them. This is a problem because we then end up with
-                        // no suggestions - CompletableFuture.allOf(...) will contain an exception if a future is cancelled,
-                        // meaning thenRun(...) does not run, which is how displaying the suggestions works...
-                        //
-                        // Because we don't control the client, we have to work around it here.
-                        if (hasCustomSuggestionsAlready && node instanceof ArgumentCommandNode) {
-                            final ArgumentCommandNode<SharedSuggestionProvider, ?> argNode = (ArgumentCommandNode<SharedSuggestionProvider, ?>) node;
-                            if (argNode.getCustomSuggestions() != null) {
-                                // Rebuild the node without the custom suggestions.
-                                rootSuggestion.addChild(this.impl$cloneArgumentCommandNodeWithoutSuggestions(argNode));
-                                continue;
-                            }
-                        } else if (node instanceof ArgumentCommandNode && ((ArgumentCommandNode<?, ?>) node).getCustomSuggestions() != null) {
-                            hasCustomSuggestionsAlready = true; // no more custom suggestions
+        final ServerPlayer e = (ServerPlayer) sourceButTyped.getEntity();
+        final Map<CommandNode<CommandSourceStack>, List<CommandNode<SharedSuggestionProvider>>> playerNodes = this.impl$playerNodeCache.get(e);
+        // Note that the value for a key can be null, this we have to check to see if it contains a key (we've processed it)
```

Comment:
Is there a way to encapsulate a "We have processed this, here is the result" without triggering the double lookup?

You trade a slightly higher memory footprint for CPU which almost always wins.

Reply:
I'll use an empty list.